src/presentation/api/controller/services.go

```diff
@@ -0,0 +1,212 @@
+			continue
+		}
+
+		rawPortBindingStr := rawPortStr + "/" + rawProtocolStr
+		rawPortBindings = append(rawPortBindings, rawPortBindingStr)
+	}
+
+	return rawPortBindings
+}
+
+// CreateInstallableService godoc
+// @Summary      CreateInstallableService
+// @Description  Install a new installable service.
+// @Tags         services
+// @Accept       json
+// @Produce      json
+// @Security     Bearer
+// @Param        createInstallableServiceDto	body dto.CreateInstallableService	true	"Only name is required.<br />If version is not provided, it will be 'lts'.<br />If portBindings is not provided, it wil be default service port bindings.<br />If autoCreateMapping is not provided, it will be 'true'."
+// @Success      201 {object} object{} "InstallableServiceCreated"
+// @Router       /v1/services/installables/ [post]
+func (controller *ServicesController) CreateInstallable(c echo.Context) error {
+	requestBody, err := apiHelper.ReadRequestBody(c)
+	if err != nil {
+		return err
+	}
+
+	rawEnvsSlice := []string{}
+	if requestBody["envs"] != nil {
+		for _, rawEnv := range requestBody["envs"].([]interface{}) {
+			rawEnvStr, err := voHelper.InterfaceToString(rawEnv)
+			if err != nil {
+				continue
+			}
+			rawEnvsSlice = append(rawEnvsSlice, rawEnvStr)
+		}
+	}
+	requestBody["envs"] = rawEnvsSlice
+
+	rawPortBindings := []string{}
+	if requestBody["portBindings"] != nil {
+		rawPortBindings = parseRawPortBindings(
+			requestBody["portBindings"].([]interface{}),
+		)
+	}
+	requestBody["portBindings"] = rawPortBindings
+
+	return apiHelper.ServiceResponseWrapper(
+		c, controller.servicesService.CreateInstallable(requestBody, true),
+	)
+}
+
+// CreateCustomService godoc
+// @Summary      CreateCustomService
+// @Description  Install a new custom service.
+// @Tags         services
+// @Accept       json
+// @Produce      json
+// @Security     Bearer
+// @Param        createCustomServiceDto	body dto.CreateCustomService	true	"name, type and startCmd is required.<br />If version is not provided, it will be 'lts'.<br />If portBindings is not provided, it wil be default service port bindings.<br />If autoCreateMapping is not provided, it will be 'true'."
+// @Success      201 {object} object{} "CustomServiceCreated"
+// @Router       /v1/services/custom/ [post]
+func (controller *ServicesController) CreateCustom(c echo.Context) error {
+	requestBody, err := apiHelper.ReadRequestBody(c)
+	if err != nil {
+		return err
+	}
+
+	rawEnvsSlice := []string{}
+	if requestBody["envs"] != nil {
+		for _, rawEnv := range requestBody["envs"].([]interface{}) {
+			rawEnvStr, err := voHelper.InterfaceToString(rawEnv)
+			if err != nil {
+				continue
+			}
+			rawEnvsSlice = append(rawEnvsSlice, rawEnvStr)
+		}
+	}
+	requestBody["envs"] = rawEnvsSlice
+
+	rawPortBindings := []string{}
+	if requestBody["portBindings"] != nil {
+		rawPortBindings = parseRawPortBindings(
+			requestBody["portBindings"].([]interface{}),
+		)
+	}
```

Comment:
portBindings may come as "3306" instead of ["3306"]. On Control, you'll see that I check that to make sure I convert that to an []interface{} instead of returning error to the user. I think you should do the same every time you asked for a []interface{}.

Reply:
I adjusted it for `portBindings`, and also did it for the `envs`.